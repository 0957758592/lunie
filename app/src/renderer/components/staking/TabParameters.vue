<template lang="pug">
  div
    div
      h3
        | Staking Parameters
        |
        i.material-icons.info-button(v-tooltip.top="paramsTooltips.description") info_outline
      .parameters__details.parameters__section
        .row
          .column
            dl.info_dl
              dt
                | Inflation Rate Change
                |
                i.material-icons.info-button(v-tooltip.top="paramsTooltips.inflation_rate_change") info_outline
              dd {{ parameters.inflation_rate_change ? parameters.inflation_rate_change : `n/a` }}
            dl.info_dl
              dt Minimum Inflation Rate
              dd {{ parameters.inflation_min ? parameters.inflation_min : `n/a` }}
            dl.info_dl
              dt Maximum Inflation Rate
              dd {{ parameters.inflation_max ? parameters.inflation_max : `n/a` }}
            dl.info_dl
              dt Goal for bonded {{ bondingDenom }}s
              dd {{ parameters.goal_bonded ? parameters.goal_bonded : `n/a` }}
          .column
            dl.info_dl
              dt
                | Unbonding time
                |
                i.material-icons.info-button(v-tooltip.top="paramsTooltips.unbonding_time") info_outline
              dd {{ parameters.unbonding_time ? parameters.unbonding_time : `n/a` }}
            dl.info_dl
              dt Max number of Validators
              dd {{ parameters.max_validators ? parameters.max_validators : `n/a` }}
            dl.info_dl
              dt Bondable coin denomination
              dd {{ parameters.bond_denom ? parameters.bond_denom : `n/a` }}
    div
      h3
        | Staking Pool
        |
        i.material-icons.info-button(v-tooltip.top="poolTooltips.description") info_outline
      .parameters__details.parameters__section
        .row
          .column
            dl.info_dl
              dt
                | Inflation
                |
                i.material-icons.info-button(v-tooltip.top="poolTooltips.inflation") info_outline
              dd {{ pool.inflation ? pool.inflation : `n/a` }}
            dl.info_dl
              dt
                | Inflation last block
                |
                i.material-icons.info-button(v-tooltip.top="poolTooltips.inflation_last_time") info_outline
              dd {{ pool.inflation_last_time ? pool.inflation_last_time : `n/a` }}
            dl.info_dl
              dt
                | Date of last commission reset
                |
                i.material-icons.info-button(v-tooltip.top="poolTooltips.date_last_commission_reset") info_outline
              dd {{ pool.date_last_commission_reset ? timeAgo(pool.date_last_commission_reset) : `n/a` }}
          .column
            dl.info_dl
              dt
                | Loose {{ bondingDenom }}
                |
                i.material-icons.info-button(v-tooltip.top="poolTooltips.loose_tokens") info_outline
              dd {{ pool.loose_tokens ? pool.loose_tokens : `n/a` }}
            dl.info_dl
              dt
                | Bonded {{ bondingDenom }}
                |
                i.material-icons.info-button(v-tooltip.top="poolTooltips.bonded_tokens") info_outline
              dd {{ pool.bonded_tokens ? pool.bonded_tokens : `n/a` }}
            dl.info_dl
              dt
                | Previous bonded shares
                |
                i.material-icons.info-button(v-tooltip.top="poolTooltips.prev_bonded_shares") info_outline
              dd {{ pool.prev_bonded_shares ? pool.prev_bonded_shares : `n/a` }}
</template>

<script>
import moment from "moment"
import { mapGetters } from "vuex"
import { TmBtn, TmListItem, TmPage, TmPart, TmToolBar } from "@tendermint/ui"
export default {
  name: `tab-staking-parameters`,
  components: {
    TmBtn,
    TmListItem,
    TmPage,
    TmPart,
    TmToolBar
  },
  data: () => ({
    paramsTooltips: {
      description: `Staking parameters define the high level settings for staking`,
      inflation_rate_change: `maximum annual change in inflation rate`,
      inflation_max: `maximum inflation rate`,
      inflation_min: `minimum inflation rate`,
      goal_bonded: `goal of percent bonded atoms`,
      unbonding_time: `time to complete an unbonding delegetion and claim bonds`,
      max_validators: `maximum number of validators for the chain`,
      bond_denom: `bondable coin denomination`
    },
    poolTooltips: {
      description: `The staking pool represents the dynamic parameters of the current state`,
      loose_tokens: `tokens which are not bonded in a validator`,
      bonded_tokens: `reserve of bonded tokens`,
      inflation_last_time: `block which the last inflation was processed`,
      inflation: `current annual inflation rate`,
      date_last_commission_reset: `UNIX timestamp for last commission accounting reset (daily)`,
      prev_bonded_shares: `last recorded bonded shares; used for fee calculations`
    }
  }),
  computed: {
    ...mapGetters([
      `bondingDenom`,
      `config`,
      `user`,
      `parameters`,
      `pool`,
      `connected`
    ])
  },
  methods: {
    timeAgo(date) {
      return moment(date, `x`).fromNow()
    },
    getStakingParameters() {
      this.$store.dispatch(`getParameters`)
    },
    getStakingPool() {
      this.$store.dispatch(`getPool`)
    }
  },
  async mounted() {
    this.getStakingParameters()
    this.getStakingPool()
  }
}
</script>
<style lang="stylus">
@require '~variables'

@media screen and (min-width: 640px)
  #parameters .tm-part-main
    display flex
    flex-flow row-reverse nowrap

    .list-items
      flex 1

.parameters

  &__details
    > .row
      > .column
        flex 1

  &__section
    background-color var(--app-fg)
    display flex
    margin-bottom 1rem
    padding 2rem
    width 100%

h3
  margin 1em auto

.info-button
  color var(--link)

.column
  display flex
  flex-flow column
  position relative

.row
  display flex
  flex-direction row
  width 100%

.info_dl
  display flex
  flex-flow column
  margin-bottom 1.5rem
  margin-right 1rem

  dt
    color var(--dim)
    font-size small
    margin-bottom 4px

  dd
    border solid 1px #31354e
    border-radius 2px
    font-size 1rem
    line-height 1rem
    padding 0.5rem

    &.info_dl__text-box
      min-height 6.91rem

@media screen and (min-width: 768px)
  padding-bottom 4rem

@media screen and (min-width: 1024px)
  .fixed-button-bar
    margin-left width-side
</style>
