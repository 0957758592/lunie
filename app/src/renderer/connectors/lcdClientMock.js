'use strict'

let state = {
  keys: [{
    name: 'MOCK_ACCOUNT',
    password: '1234567890',
    address: 'DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B'
  }],
  accounts: {
    DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B: {
      coins: [{
        denom: 'mycoin',
        amount: 1000
      }]
    }
  },
  nonces: { DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B: 0 },
  txs: [{
    tx: {
      hash: 'x',
      inputs: [
        {
          coins: [{
            denom: 'jbcoins',
            amount: 1234
          }],
          sender: makeAddress()
        }
      ],
      outputs: [
        {
          coins: [{
            denom: 'jbcoins',
            amount: 1234
          }],
          receiver: 'DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B'
        }
      ]
    },
    height: 1
  }, {
    tx: {
      hash: 'y',
      inputs: [
        {
          coins: [{
            denom: 'fabocoins',
            amount: 1234
          }],
          sender: 'DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B'
        }
      ],
      outputs: [
        {
          coins: [{
            denom: 'fabocoins',
            amount: 1234
          }],
          receiver: makeAddress()
        }
      ]
    },
    height: 150
  }],
  stake: {
    DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B: {
      '7A9D783CE542B23FA23DC7F101460879861205772606B4C3FAEAFBEDFB00E7BD': {
        PubKey: { data: '7A9D783CE542B23FA23DC7F101460879861205772606B4C3FAEAFBEDFB00E7BD' },
        Shares: 5
      }
    }
  },
  delegates: [
    {
      'address': '70705055A9FA5901735D0C3F0954501DDE667327',
      'pub_key': {
        'type': 'ed25519',
        'data': '88564A32500A120AA72CEFBCF5462E078E5DDB70B6431F59F778A8DC4DA719A4'
      },
      'voting_power': 14,
      'accum': 585,
      description: {
        description: 'Mr Mounty',
        moniker: 'mr_mounty',
        country: 'Canada'
      }
    },
    {
      'address': '760ACDE75EFC3DD0E4B2A6A3B96D91C05349EA31',
      'pub_key': {
        'type': 'ed25519',
        'data': '7A9D783CE542B23FA23DC7F101460879861205772606B4C3FAEAFBEDFB00E7BD'
      },
      'voting_power': 32,
      'accum': -1107,
      description: {
        description: 'Good Guy Greg',
        moniker: 'good_greg',
        country: 'USA'
      }
    },
    {
      'address': '77C26DF82654C5A5DDE5C6B7B27F3F06E9C223C0',
      'pub_key': {
        'type': 'ed25519',
        'data': '651E7B12B3C7234FB82B4417C59DCE30E4EA28F06AD0ACAEDFF05F013E463F10'
      },
      'voting_power': 19,
      'accum': 539,
      description: {
        description: 'Herr Schmidt',
        moniker: 'herr_schmidt',
        country: 'DE'
      }
    }
  ]
}

module.exports = {
  async lcdConnected () { return true },
  async sign ({ name, password, tx }) {
    return {
      'type': 'sigs/one',
      'data': {
        'tx': tx,
        'signature': {
          'Sig': {
            'type': 'ed25519',
            'data': '4631B03666EDF3D95E67A0700AD93741728E9466E51A8684F8637B935BB9A296ACE66256F2EDEC4F469FB9A940A4C1D30BB9170B5CA5292A6570A9F836DB7609'
          },
          'Pubkey': {
            'type': 'ed25519',
            'data': '81B11E717789600CC192B26F452A983DF13B985EE75ABD9DD9E68D7BA007A958'
          }
        }
      }
    }
  },
  async postTx (signedTx) {
    let outerTx = signedTx.data.tx.data.tx.data.tx
    let address = outerTx.data.signers[0].addr
    state.txs.push(outerTx.data.tx)
    state.nonces[address]++

    return {
      'check_tx': {
        'code': 0,
        'data': '',
        'log': '',
        'gas': '0',
        'fee': '0'
      },
      'deliver_tx': {
        'code': 0,
        'data': '',
        'log': '',
        'tags': []
      },
      'hash': '999ADECC2DE8C3AC2FD4F45E5E1081747BBE504A',
      'height': 0
    }
  },

  // keys
  async generateKey ({ name, password }) {
    let key = {
      name, password, address: makeAddress()
    }
    state.keys.push(key)
    return key
  },
  async listKeys () {
    return state.keys.map(k => ({
      name: k.name,
      address: k.address
    }))
  },
  async getKey (name) {
    return state.keys.find(k => k.name === name)
  },
  async updateKey (account, { name, password, new_passphrase }) { // eslint-disable-line camelcase
    let key = state.keys.find(k => k.name === name)
    if (key.password !== password) {
      throw new Error('Passwords do not match')
    }
    key.password = new_passphrase // eslint-disable-line camelcase
  },
  // axios handles DELETE requests different then other requests, we have to but the body in a config object with the prop data
  async deleteKey (account, { name, password }) {
    let key = state.keys.find(k => k.name === name)
    key
    if (key.password !== password) {
      throw new Error('Passwords do not match')
    }
    state.keys = state.keys.filter(k => k.name === name)
  },
  async recoverKey ({ name, password, seed_phrase }) { // eslint-disable-line camelcase
    let key = {
      name, password, address: makeAddress()
    }
    state.keys.push(key)
    return key
  },

  // coins
  async buildSend ({ sequence, from: { addr: fromAddr }, to: { addr: toAddr }, fees, amount }) {
    return {
      'type': 'sigs/one',
      'data': {
        'tx': {
          'type': 'chain/tx',
          'data': {
            'chain_id': 'mock-chain',
            'expires_at': 0,
            'tx': {
              'type': 'nonce',
              'data': {
                'sequence': sequence,
                'signers': [
                  {
                    'chain': '',
                    'app': 'sigs',
                    'addr': fromAddr
                  }
                ],
                'tx': {
                  'type': 'coin/send',
                  'data': {
                    'inputs': [
                      {
                        'address': {
                          'chain': '',
                          'app': 'sigs',
                          'addr': fromAddr
                        },
                        'coins': amount
                      }
                    ],
                    'outputs': [
                      {
                        'address': {
                          'chain': '',
                          'app': 'sigs',
                          'addr': toAddr
                        },
                        'coins': amount
                      }
                    ]
                  }
                }
              }
            }
          }
        },
        'signature': {
          'Sig': null,
          'Pubkey': null
        }
      }
    }
  },
  async queryAccount (address) {
    let account = state.accounts[address]
    if (account) {
      return { data: state.accounts[address] }
    }
    return { data: { coins: [] } }
  },
  async coinTxs (address) {
    return state.txs.filter(tx => {
      return tx.tx.inputs.find(input => input.sender === address) || tx.tx.outputs.find(output => output.receiver === address)
    })
  },

  // nonce
  async queryNonce (address) {
    return state.nonces[address]
  },

  // staking
  async candidate (pubkey) {
    return { data: state.delegates.find(delegate => delegate.pub_key.data === pubkey) }
  },
  async candidates () {
    return { data: state.delegates.map(({ pub_key }) => pub_key) } // eslint-disable-line camelcase
  },
  async buildDelegate ({ sequence, from, amount, pubkey }) {
    return {
      'type': 'sigs/one',
      'data': {
        'tx': {
          'type': 'chain/tx',
          'data': {
            'chain_id': 'mock-chain',
            'expires_at': 0,
            'tx': {
              'type': 'nonce',
              'data': {
                'sequence': sequence,
                'signers': [
                  {
                    'chain': '',
                    'app': 'sigs',
                    'addr': from.addr
                  }
                ],
                'tx': {
                  'type': 'stake/delegate',
                  'data': {
                    'pub_key': pubkey,
                    'amount': amount
                  }
                }
              }
            }
          }
        },
        'signature': {
          'Sig': null,
          'Pubkey': null
        }
      }
    }
  },
  async buildUnbond () {
    throw Error('Not implemented')
  },
  async bondingsByDelegator ([address, delegatePubkey]) {
    let stake = state.stake[address]
    if (stake && stake[delegatePubkey]) {
      return { data: stake[delegatePubkey] }
    }
    return {
      data: {
        PubKey: { data: delegatePubkey },
        Shares: 0
      }
    }
  }
}

function makeAddress () {
  var text = ''
  var possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

  for (var i = 0; i < 40; i++) { text += possible.charAt(Math.floor(Math.random() * possible.length)) }

  return text
}
