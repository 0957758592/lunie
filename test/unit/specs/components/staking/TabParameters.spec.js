import moment from "moment"
import Vuelidate from "vuelidate"
import setup from "../../../helpers/vuex-setup"
import htmlBeautify from "html-beautify"
import TabParameters from "renderer/components/staking/TabParameters"
import lcdClientMock from "renderer/connectors/lcdClientMock.js"

let pool = lcdClientMock.state.pool
let parameters = lcdClientMock.state.parameters

describe(`TabParameters`, () => {
  let wrapper, store
  let { mount, localVue } = setup()
  localVue.use(Vuelidate)

  const $store = {
    commit: jest.fn(),
    dispatch: jest.fn(),
    getters: {
      pool,
      parameters,
      totalAtoms: 100,
      user: { atoms: 42 }
    }
  }

  beforeEach(() => {
    let instance = mount(TabParameters, {
      localVue,
      doBefore: ({ store }) => {
        store.commit(`setPool`, pool)
        store.commit(`setParameters`, parameters)
      },
      $store
    })
    wrapper = instance.wrapper
    store = instance.store
    store.commit(`setConnected`, true)
    wrapper.update()
  })

  it(`has the expected html structure`, async () => {
    await wrapper.vm.$nextTick()
    wrapper.update()
    expect(htmlBeautify(wrapper.html())).toMatchSnapshot()
  })

  it(`shows the staking parameters and pool`, () => {
    expect(store.state.parameters).toEqual(parameters)
    expect(store.state.pool).toEqual(pool)
  })

  it(`displays unbonding period in days`, () => {
    let days = moment.duration(parameters.unbonding_time).asDays()
    expect(wrapper.vm.unbondingTimeInDays).toEqual(days)
  })
})
