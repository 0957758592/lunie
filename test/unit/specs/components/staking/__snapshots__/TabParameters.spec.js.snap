// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabParameters has the expected html structure 1`] = `
"<div>
   <div>
      <h3>
         Staking Parameters<i class=\\"material-icons info-button\\">info_outline</i>
      </h3>
      <div class=\\"parameters__details parameters__section\\">
         <div class=\\"row\\">
            <div class=\\"column\\">
               <dl class=\\"info_dl\\">
                  <dt>
                     Inflation Rate Change<i class=\\"material-icons info-button\\">info_outline</i>
                  </dt>
                  <dd>
                     n/a
                  </dd>
</dl>
               <dl class=\\"info_dl\\">
                  <dt>
                     Minimum Inflation Rate
                  </dt>
                  <dd>
                     n/a
                  </dd>
</dl>
               <dl class=\\"info_dl\\">
                  <dt>
                     Maximum Inflation Rate
                  </dt>
                  <dd>
                     n/a
                  </dd>
</dl>
               <dl class=\\"info_dl\\">
                  <dt>
                     Goal for bonded Steaks
                  </dt>
                  <dd>
                     n/a
                  </dd>
</dl>
</div>
<div class=\\"column\\">
   <dl class=\\"info_dl\\">
      <dt>
         Unbonding time<i class=\\"material-icons info-button\\">info_outline</i>
      </dt>
      <dd>
         n/a
      </dd>
</dl>
   <dl class=\\"info_dl\\">
      <dt>
         Max number of Validators
      </dt>
      <dd>
         n/a
      </dd>
</dl>
   <dl class=\\"info_dl\\">
      <dt>
         Bondable coin denomination
      </dt>
      <dd>
         n/a
      </dd>
</dl>
</div>
</div>
</div>
</div>
<div>
   <h3>
      Staking Pool<i class=\\"material-icons info-button\\">info_outline</i>
   </h3>
   <div class=\\"parameters__details parameters__section\\">
      <div class=\\"row\\">
         <div class=\\"column\\">
            <dl class=\\"info_dl\\">
               <dt>
                  Inflation<i class=\\"material-icons info-button\\">info_outline</i>
               </dt>
               <dd>
                  n/a
               </dd>
</dl>
            <dl class=\\"info_dl\\">
               <dt>
                  Inflation last block<i class=\\"material-icons info-button\\">info_outline</i>
               </dt>
               <dd>
                  n/a
               </dd>
</dl>
            <dl class=\\"info_dl\\">
               <dt>
                  Date of last commission reset<i class=\\"material-icons info-button\\">info_outline</i>
               </dt>
               <dd>
                  n/a
               </dd>
</dl>
</div>
<div class=\\"column\\">
   <dl class=\\"info_dl\\">
      <dt>
         Loose Steak<i class=\\"material-icons info-button\\">info_outline</i>
      </dt>
      <dd>
         n/a
      </dd>
</dl>
   <dl class=\\"info_dl\\">
      <dt>
         Bonded Steak<i class=\\"material-icons info-button\\">info_outline</i>
      </dt>
      <dd>
         n/a
      </dd>
</dl>
   <dl class=\\"info_dl\\">
      <dt>
         Previous bonded shares<i class=\\"material-icons info-button\\">info_outline</i>
      </dt>
      <dd>
         n/a
      </dd>
</dl>
</div>
</div>
</div>
</div>
</div>"
`;
